samza-api/src/main/java/org/apache/samza/operators/StreamGraph.java

```diff
@@ -0,0 +1,118 @@
+   * @return   the input {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> MessageStream<M> createInStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
+
+  /**
+   * Method to add an output {@link MessageStream} from the system
+   *
+   * @param streamSpec  the {@link StreamSpec} describing the physical characteristics of the output {@link MessageStream}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the output {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the output {@link MessageStream}
+   * @param <K>  the type of key in the output message
+   * @param <V>  the type of message in the output message
+   * @param <M>  the type of {@link MessageEnvelope} in the output {@link MessageStream}
+   * @return   the output {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createOutStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
+
+  /**
+   * Method to add an intermediate {@link MessageStream} from the system
+   *
+   * @param streamSpec  the {@link StreamSpec} describing the physical characteristics of the intermediate {@link MessageStream}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the intermediate {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the intermediate {@link MessageStream}
+   * @param <K>  the type of key in the intermediate message
+   * @param <V>  the type of message in the intermediate message
+   * @param <M>  the type of {@link MessageEnvelope} in the intermediate {@link MessageStream}
+   * @return   the intermediate {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createIntStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
+
+  /**
+   * Method to get the input {@link MessageStream}s
+   *
+   * @return the input {@link MessageStream}
+   */
+  Map<StreamSpec, MessageStream> getInStreams();
+
+  /**
+   * Method to get the {@link OutputStream}s
+   *
+   * @return  the map of all {@link OutputStream}s
+   */
+  Map<StreamSpec, OutputStream> getOutStreams();
+
+  /**
+   * Method to set the {@link ContextManager} for this {@link StreamGraph}
+   *
+   * @param manager  the {@link ContextManager} object
+   * @return  this {@link StreamGraph} object
+   */
+  StreamGraph withContextManager(ContextManager manager);
+
+  String GRAPH_CONFIG = "job.stream.graph.impl.class";
+  String DEFAULT_GRAPH_IMPL_CLASS = "org.apache.samza.operators.StreamGraphImpl";
+
+  /**
+   * Static method to instantiate the implementation class of {@link StreamGraph}.
+   *
+   * @param config  the {@link Config} object for this job
+   * @return  the {@link StreamGraph} object created
+   */
+  static StreamGraph fromConfig(Config config) {
```

Comment:
Yeah. Thanks for the suggestion. I have made the change to this one. Now, I have removed this fromConfig() method and make the change to StreamGraphFactory ==> StreamGraphBuilder.init(graph, config), in which graph is created by the framework and passed into the StreamGraphBuilder's init() method. This also removed one questionable line in the user example code that user needs to call StreamGraph.fromConfig(config) to actually create the StreamGraphImpl object, which user has no public method to instantiate the object. Now, the StreamGraphImpl object is only instantiated by the framework and user only modify it via init() method.